Vyper Compiler Notes (#1635)
vyperlang/vyper: docs/compiler-notes.rst

Compiler Notes
**************

Vyper Compilation Stages
========================

- Python AST
- Python AST to Vyper AST
- Vyper AST to LLL
- LLL to Bytecode

Differences from Python
=======================

Introduction
~~~~~~~~~~~~

As Vyper targets Smart Contract development - there are some notable differences. This section
aims to highlight the important parts.

Types
~~~~~

> One of the key difference between Python and Vyper is that Python usese dynamic typing; and Vyper

Reviewer: Typo: "usese".

> One of the key difference between Python and Vyper is that Python usese dynamic typing; and Vyper

Reviewer 2: ```suggestion
One of the key difference between Python and Vyper is that Python uses dynamic typing; and Vyper
```

> uses static Typing. Type in Vyper are declared using Python Annotations e.g. `a: auint256`.

Reviewer 2: ```suggestion
uses static Typing. Type in Vyper are declared using Python Annotations e.g. `a: uint256`.
```

When declaring a global a `public()` can create a getter function for the value. e.g.
`a: public(uint256)`.

*Events / Logging*

Events are defined using the Annotation of a ast.Call with a name of `event`.

```
Transfer: event({_from: indexed(address), _to: indexed(address), _value: uint256})
```

The `event` takes a single argument of `ast.Dict` with optional `indexed` around the specific types -
which indicates wether a topic is index or not (see the LOG opcode).

*Interfaces / Interface Definitions *

Interfaces can be imported from external files as well defined locally within a contract using the
`contract` keyword.

Unsupported python statements and expressions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- `yield`
- `async`
- `await`
- `async for`
- `try`
- `with`
- `while`
- `global`
- `nonlocal`
- `lambda` functions
- bitwise shifting operators
  - The use of bitwise shiftiing operators is not supported, instead `bitewise_*` functions exist.
- shifting operators not supported. Use `shift` insetad.
- floor division operator `//` is not supported. Use `floor` and `ceil` instead, to make rounding
  operations clear.

Statements
~~~~~~~~~~

*Simple Statements*

> **Airthmetic Conversions**

Reviewer 2: ```suggestion
**Arithmetic Conversions**
```

> Vyper enforces airthmetic conversions, to be explicit vs. python which uses implicit type conversion.

Reviewer 2: ```suggestion
Vyper enforces arithmetic conversions, to be explicit vs. python which uses implicit type conversion.
```

This was chosen specifically, to make the conversion steps clearer. The expression `1 * 1.0` will
be disallowed, instead `convert(1, decimal) * 1.0`.

**Identifiers**

Vyper ensures variables names can only consist of alphanumeric characters, with `_`. The specific
regex tested against is `^[_a-zA-Z][a-zA-Z0-9_]*$`.

**Literals**

Vyper has support for integer, decimal, string and byte python literals.
- If to Literals are found in an expression they are constant folded.
  - `10*10` will be converted to `100`
  - However `10*10*a` will not be converted.
- See above artihmetic onversion rules.
- There are limitations to byte & string literals based on their usage.
  - e.g. A string literal of size 100 can't be assigned or return to an annoted size of 32.
- To distinguish between `int128` and `uint256` literals Vyper uses the following ruleset.

>   - If literal int value is above MAXNUM (2**127 - 1), literal is a uint256.

Reviewer 2: Would this be a correct way of stating this?

```suggestion
  - If literal int value is above `MAXNUM` (`2**127 - 1`), the literal must be assigned to an uint256.
```

  - If literal int value is under MINNUM (-2**127), literal is invalid.
  - If assignment of literal to a known type (could be uint256 or int128), use appropriate type.
  - Consequently slightly confusing error message of `Cannot assign uint256 to int128` does occur,

>     but seemed to be a reasonable trade off (don't have to convert(xxxx, uint256) on assingment).

Reviewer 2: ```suggestion
    but seemed to be a reasonable trade off (don't have to convert(xxxx, uint256) on assignment).
```

**Comparisons / Boolean Operations*

- Comparisons

>   - As long as all members of are of the same type -comparisons are supported.

Reviewer 2: ```suggestion
  - As long as all members of are of the same type, comparisons are supported.
```

- Boolean `and`, `or` operations are supported.

>     - Currently not short-circuited (VIP pending)

Reviewer 2: Link to VIP?

** Membership **

- Array membership e.g. `assert a in self.owners` is supported.

** del Delete Statement **

Vyper does not support `del ..` to clear values from storage use `clear()` instead.

** Import Statements **

Import statements are supported for use with import interfaces, and do not import functions.

** Assert Statements **

> `assert` is supported and uses the `REVERT` opcode. To `INVALID` opcode, use assert <expre>, UNREACHABLE

Reviewer 2: ```suggestion
`assert` is supported and uses the `REVERT` opcode. To use `INVALID` opcode, use assert <expre>, UNREACHABLE
```

** Raise Statements **

`raise` is supported, uses the `REVERT` opcode to throw an exception (with reason string smaller 
than 32 bytes)

** Return Statements **

- `return` statements are supported.
- There is a checker in place to ensure return statements are not missing in branches.

> *Compound Statements*

Reviewer 2: Missing stuff here

**Functions**

> - Vyper support default arguments for both public and private functions.

Reviewer 2: What happens with default arguments in private functions? Do we have test cases for this?
